File: api/services/events.py
"""
Server-Sent Events (SSE) for real-time updates
"""
import asyncio
import json
from typing import AsyncGenerator
from datetime import datetime


class EventBroadcaster:
    """Simple event broadcaster for SSE"""

    def __init__(self):
        self._subscribers: list[asyncio.Queue] = []

    async def subscribe(self) -> AsyncGenerator[str, None]:
        """Subscribe to events"""
        queue: asyncio.Queue = asyncio.Queue()
        self._subscribers.append(queue)
        try:
            while True:
                data = await queue.get()
                yield data
        finally:
            self._subscribers.remove(queue)

    async def broadcast(self, event_type: str, data: dict = None):
        """Broadcast an event to all subscribers"""
        event = {
            "type": event_type,
            "data": data or {},
            "timestamp": datetime.utcnow().isoformat()
        }
        message = f"data: {json.dumps(event)}\n\n"

        for queue in self._subscribers:
            await queue.put(message)

    @property
    def subscriber_count(self) -> int:
        return len(self._subscribers)


# Global event broadcaster instance
library_events = EventBroadcaster()

# Migration event broadcaster instance
migration_events = EventBroadcaster()

# Subtitle event broadcaster instance
subtitle_events = EventBroadcaster()

# Per-session share sync event broadcasters
# Key: share session token, Value: EventBroadcaster
share_sync_events: dict[str, EventBroadcaster] = {}


def get_share_broadcaster(token: str) -> EventBroadcaster:
    """Get or create a broadcaster for a share session."""
    if token not in share_sync_events:
        share_sync_events[token] = EventBroadcaster()
    return share_sync_events[token]


# Event types
class EventType:
    IMAGE_ADDED = "image_added"
    IMAGE_UPDATED = "image_updated"
    IMAGE_DELETED = "image_deleted"
    TASK_COMPLETED = "task_completed"


class MigrationEventType:
    STARTED = "migration_started"
    PROGRESS = "migration_progress"
    COMPLETED = "migration_completed"
    ERROR = "migration_error"


class SubtitleEventType:
    STARTED = "subtitle_started"
    CUE = "subtitle_cue"
    PROGRESS = "subtitle_progress"
    COMPLETED = "subtitle_completed"
    ERROR = "subtitle_error"


# Cast event broadcaster instance
cast_events = EventBroadcaster()


class CastEventType:
    STATUS = "cast_status"
    CONNECTED = "cast_connected"
    DISCONNECTED = "cast_disconnected"
    ERROR = "cast_error"

File: src-tauri/src/services/events.rs
use std::convert::Infallible;
use std::sync::Arc;

use chrono::Utc;
use serde_json::{json, Value};
use tokio::sync::broadcast;

/// Channel capacity per broadcaster.
const CHANNEL_CAPACITY: usize = 256;

// ─── Event type constants ────────────────────────────────────────────────────

pub mod event_type {
    pub const IMAGE_ADDED: &str = "image_added";
    pub const IMAGE_UPDATED: &str = "image_updated";
    pub const IMAGE_DELETED: &str = "image_deleted";
    pub const TASK_COMPLETED: &str = "task_completed";
}

pub mod migration_event_type {
    pub const STARTED: &str = "migration_started";
    pub const PROGRESS: &str = "migration_progress";
    pub const COMPLETED: &str = "migration_completed";
    pub const ERROR: &str = "migration_error";
}

// ─── EventBroadcaster ────────────────────────────────────────────────────────

/// Simple pub/sub event broadcaster backed by `tokio::sync::broadcast`.
///
/// Subscribers receive formatted SSE messages (`data: {...}\n\n`).
/// Slow subscribers that fall behind the capacity will skip missed events.
#[derive(Clone)]
pub struct EventBroadcaster {
    tx: broadcast::Sender<String>,
}

impl EventBroadcaster {
    pub fn new() -> Self {
        let (tx, _) = broadcast::channel(CHANNEL_CAPACITY);
        Self { tx }
    }

    /// Broadcast an event to all current subscribers.
    pub fn broadcast(&self, event_type: &str, data: Value) {
        let event = json!({
            "type": event_type,
            "data": data,
            "timestamp": Utc::now().to_rfc3339()
        });
        let message = format!("data: {}\n\n", event);
        // Ignore error if no subscribers
        let _ = self.tx.send(message);
    }

    /// Subscribe and get a receiver that yields SSE-formatted strings.
    pub fn subscribe(&self) -> broadcast::Receiver<String> {
        self.tx.subscribe()
    }

    /// Create an SSE-compatible stream for axum.
    pub fn sse_stream(
        &self,
    ) -> impl futures_core::Stream<Item = Result<axum::response::sse::Event, Infallible>> {
        let mut rx = self.subscribe();

        async_stream::stream! {
            // Initial connection message
            let connected = json!({"type": "connected", "timestamp": Utc::now().to_rfc3339()});
            yield Ok(axum::response::sse::Event::default().data(connected.to_string()));

            loop {
                match rx.recv().await {
                    Ok(msg) => {
                        // The message is already formatted as "data: {...}\n\n"
                        // but axum SSE wraps it, so extract the JSON part
                        let json_part = msg
                            .strip_prefix("data: ")
                            .and_then(|s| s.strip_suffix("\n\n"))
                            .unwrap_or(&msg);
                        yield Ok(axum::response::sse::Event::default().data(json_part.to_string()));
                    }
                    Err(broadcast::error::RecvError::Lagged(n)) => {
                        log::warn!("SSE subscriber lagged by {} events", n);
                        // Continue receiving — subscriber just missed some events
                    }
                    Err(broadcast::error::RecvError::Closed) => {
                        break;
                    }
                }
            }
        }
    }

    /// Number of active subscribers (approximate — includes lagged receivers).
    pub fn subscriber_count(&self) -> usize {
        self.tx.receiver_count()
    }
}

impl Default for EventBroadcaster {
    fn default() -> Self {
        Self::new()
    }
}

// ─── Global instances ────────────────────────────────────────────────────────

/// Holds all global event broadcaster instances.
pub struct Events {
    pub library: EventBroadcaster,
    pub migration: EventBroadcaster,
}

impl Events {
    pub fn new() -> Self {
        Self {
            library: EventBroadcaster::new(),
            migration: EventBroadcaster::new(),
        }
    }
}

impl Default for Events {
    fn default() -> Self {
        Self::new()
    }
}

/// Thread-safe shared reference to global events.
pub type SharedEvents = Arc<Events>;

/// Create a new shared events instance.
pub fn create_events() -> SharedEvents {
    Arc::new(Events::new())
}
